### DISCUSSION TITLE
using should provide access to contained namespaces as well as types

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#3968] [Language Ideas]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/1201345/using-should-provide-access-to-contained-namespace.html)._

---
<p>Allow using to do a partial namespace qualification.  Currently, you cannot do something like</p><pre>using System;
IO. Stream myStream;</pre> <p>you have to either do the fully qualified System.IO.Stream in the variable declaration or add an imports with the full System.IO namespace so that you can use Stream as the type name.  The former leads to long, cumbersome, namespace qualified names everywhere, and the latter results in a huge blizzard of using statements everywhere, and unwieldy long intellisense lists.  There are many cases where one using for one outer namespace will do, and using a single inner namespace name to qualify a type is conveniently terse enough.  VB has had this since version 1.0, and it was never a problem.  In the case of ambiguity, a fully qualified name would still be required.</p>

---
#### Mila Zhou[MSFT] on 9/27/2020, 00:32 AM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by canton7** [comment]
Note that you can achieve the same by doing:

```cs
using IO = System.IO;
```

**Comment by scottdorman** [comment]
> huge blizzard of using statements everywhere

Good program organization puts all of the `using` directives at the top of the file and every tool (that I'm aware of at least) which allows you to organize your code, does anything with the `using` directives, etc. puts them at the top.

> unwieldy long intellisense lists.

This is a Visual Studio thing, not a language thing. (Yes, the list of things to display in IntelliSense comes from the language service, but how it's displayed is Visual Studio's responsibility). Visual display of something in the editor isn't really going to be a reason for a language change. The list supports filtering by member type (using the tabs at the bottom of the list).

**Reply by canton7** [reply]
It's also worth adding that VS Preview will suggest types in different namespaces, so your usings make very little difference.

**Reply by yaakov-h** [reply]
I don't think that's just a Preview thing anymore.

**Reply by Joe4evr** [reply]
Yea, it's been in Stable for a good while now.
